### DISCUSSION TITLE
Root level command naming convention

### DISCUSSION THREAD

**Discussion by schmidsi** [scaffold-eth/scaffold-eth-2#977] [Ideas]
When running `yarn run` I see different commands, some with `<package>:` prefix and some without. Also, some subgraph specific commands seem to follow the pattern `node-<command>`.

```bash
➜  savp-prototype git:(main) yarn run
➤ YN0000: abi-copy              'yarn workspace @se-2/subgraph abi-copy'
➤ YN0000: account               'yarn workspace @se-2/hardhat account'
➤ YN0000: chain                 'yarn workspace @se-2/hardhat chain'
➤ YN0000: clean-node            'yarn workspace @se-2/subgraph clean-node'
➤ YN0000: codegen               'yarn workspace @se-2/subgraph codegen'
➤ YN0000: compile               'yarn workspace @se-2/hardhat compile'
➤ YN0000: deploy                'yarn workspace @se-2/hardhat deploy'
➤ YN0000: flatten               'yarn workspace @se-2/hardhat flatten'
➤ YN0000: fork                  'yarn workspace @se-2/hardhat fork'
➤ YN0000: format                'yarn next:format && yarn hardhat:format'
➤ YN0000: generate              'yarn workspace @se-2/hardhat generate'
➤ YN0000: graph                 'yarn workspace @se-2/subgraph graph'
➤ YN0000: hardhat-verify        'yarn workspace @se-2/hardhat hardhat-verify'
➤ YN0000: hardhat:format        'yarn workspace @se-2/hardhat format'
➤ YN0000: hardhat:lint          'yarn workspace @se-2/hardhat lint'
➤ YN0000: hardhat:lint-staged   'yarn workspace @se-2/hardhat lint-staged'
➤ YN0000: hardhat:test          'yarn workspace @se-2/hardhat test'
➤ YN0000: postinstall           'husky install'
➤ YN0000: local-create          'yarn workspace @se-2/subgraph local-create'
➤ YN0000: local-ship            'yarn workspace @se-2/subgraph local-ship'
➤ YN0000: next:build            'yarn workspace @se-2/nextjs build'
➤ YN0000: next:check-types      'yarn workspace @se-2/nextjs check-types'
➤ YN0000: next:format           'yarn workspace @se-2/nextjs format'
➤ YN0000: next:lint             'yarn workspace @se-2/nextjs lint'
➤ YN0000: next:serve            'yarn workspace @se-2/nextjs serve'
➤ YN0000: precommit             'lint-staged'
➤ YN0000: run-node              'yarn workspace @se-2/subgraph run-node'
➤ YN0000: start                 'yarn workspace @se-2/nextjs dev'
➤ YN0000: stop-node             'yarn workspace @se-2/subgraph stop-node'
➤ YN0000: subgraph:test         'yarn workspace @se-2/subgraph test'
➤ YN0000: test                  'yarn hardhat:test'
➤ YN0000: vercel                'yarn workspace @se-2/nextjs vercel'
➤ YN0000: vercel:yolo           'yarn workspace @se-2/nextjs vercel:yolo'
➤ YN0000: verify                'yarn workspace @se-2/hardhat verify'
➤ YN0000: Done in 0s 3ms
```

I suggest to come up with a unified naming convention for these top-level commands:
```
<package>:<command>
```

Exceptions are commands that span across multiple packages or are reserved words:
- Reserved words: `precommit`,  `postinstall`, etc.
- Multi package: `test` (should also run subgraph and nextjs tests if there are any), `dev` (should spin up a local dev environment, e.g. running `hardhat:chain`, `hardhat:deploy`, `nextjs:dev` sequentially), etc.

Thoughts, objections? 

Maybe it makes sense to have a third category of exceptions which would be "getting started or happy path commands"

**Comment by carletex** [comment]
> `<package>:<command>`

I think this is a great idea. I'm definitely down for having a standardized naming convention

> dev (should spin up a local dev environment, e.g. running hardhat:chain, hardhat:deploy, nextjs:dev sequentially), etc.

This is tricky since some commands don't finish (e.g chain). We tried something like this recently (using https://www.npmjs.com/package/stmux + https://www.npmjs.com/package/wait-on): it was great, but poor Windows support for people not using WSL :(

e.g.

```
"dev:deploy": "wait-on http://127.0.0.1:8545 && yarn deploy",
"dev:next": "wait-on http://127.0.0.1:42069 && yarn start",
"dev": "stmux -t \"Scaffold-ETH 2\" -w always -e ERROR -m beep,system -M [ [ \"yarn chain\" .. \"yarn dev:deploy\" ] : [ \"yarn dev:next\" .. \"$SHELL || bash || zsh || powershell.exe || cmd.exe\" ] ]"
```

> Maybe it makes sense to have a third category of exceptions which would be "getting started or happy path commands"

Yes, I think the current commands `chain`, `deploy`, `start` should still exist (just being aliases to `hardhat/foundry:chain`, etc)

**Reply by schmidsi** [reply]
I see your point with the `dev` problems. I once tried to separate those steps into `dev`, which spins up all dev-processes: chain, next.js and graph-node and/or others. And then have a `quickstart` command that bootstraps everything: deploy, some initial transactions, deploy subgraph, etc.

See my old hackathon-starterkit:
- [`dev`](https://github.com/schmidsi/hackathon-starterkit/blob/main/package.json#L3)
- [`quickstart`](https://github.com/schmidsi/hackathon-starterkit/blob/main/package.json#L5)

Did not try on Windows though and it uses Docker.

I'll do a pull-request with the agreed upon naming and skip the dev command for now.

**Comment by schmidsi** [comment]
See pull request here: https://github.com/scaffold-eth/scaffold-eth-2/pull/1006
